File: 2016/1/solution.py
import os
from typing import List, Set, Tuple

def parse(puzzle_input: str) -> List[Tuple[str, int]]:
  instructions: List[Tuple[str, int]] = []
  for line in puzzle_input.split(', '):
    instructions.append((line[0], int(line[1:])))
  return instructions

def part1(instructions: List[Tuple[str, int]]) -> int:
  pos = [0, 0]
  direc = [0, 1]
  for instruction in instructions:
    if instruction[0] == 'L':
      direc[0], direc[1] = -direc[1], direc[0]
    else:
      direc[0], direc[1] = direc[1], -direc[0]
    pos[0] += instruction[1] * direc[0]
    pos[1] += instruction[1] * direc[1]
  return abs(pos[0]) + abs(pos[1])

def part2(instructions: List[Tuple[str, int]]) -> int:
  pos = [0, 0]
  direc = [0, 1]
  visited: Set[Tuple[int, ...]] = set()
  for instruction in instructions:
    if instruction[0] == 'L':
      direc[0], direc[1] = -direc[1], direc[0]
    else:
      direc[0], direc[1] = direc[1], -direc[0]
    for _ in range(instruction[1]):
      pos[0] += direc[0]
      pos[1] += direc[1]
      if tuple(pos) in visited:
        return abs(pos[0]) + abs(pos[1])
      visited.add(tuple(pos))
  return -1

if __name__ == '__main__':
  with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r') as file:
    puzzle_input = file.readline().strip()
  instructions = parse(puzzle_input)
  print(part1(instructions))
  print(part2(instructions))

File: 2016/1/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), ", "))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

type Instruction struct {
	turn     byte
	distance int
}

func parseInput(data []string) []*Instruction {
	instructions := make([]*Instruction, len(data))
	for i, step := range data {
		turn := step[0]
		distance, err := strconv.Atoi(step[1:])
		if err != nil {
			panic(err)
		}
		instructions[i] = &Instruction{turn, distance}
	}
	return instructions
}

func part1(puzzleInput []*Instruction) int {
	pos := [2]int{0, 0}
	dir := [2]int{0, 1}
	for _, instruction := range puzzleInput {
		if instruction.turn == 'L' {
			dir[0], dir[1] = -dir[1], dir[0]
		} else {
			dir[0], dir[1] = dir[1], -dir[0]
		}
		pos[0] += dir[0] * instruction.distance
		pos[1] += dir[1] * instruction.distance
	}
	return utils.AbsInt(pos[0]) + utils.AbsInt(pos[1])
}

func part2(puzzleInput []*Instruction) int {
	pos := [2]int{0, 0}
	dir := [2]int{0, 1}
	history := make(map[[2]int]struct{})
	history[[2]int{0, 0}] = struct{}{}
	for _, instruction := range puzzleInput {
		if instruction.turn == 'L' {
			dir[0], dir[1] = -dir[1], dir[0]
		} else {
			dir[0], dir[1] = dir[1], -dir[0]
		}
		for range instruction.distance {
			pos[0] += dir[0]
			pos[1] += dir[1]
			if _, exists := history[pos]; exists {
				return utils.AbsInt(pos[0]) + utils.AbsInt(pos[1])
			}
			history[pos] = struct{}{}
		}
	}
	return -1
}
